### PR TITLE
Merge target into head

### PR DESCRIPTION
**Pull request by xvjiarui**
This PR merged `anchor_target` function into `AnchorHead`, and `bbox_target` into `BBoxHead`.

### CODE REVIEW

File: mmdet/core/anchor/anchor_utils.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import torch
```
Comment (ZwwWayne) [inline mmdet/core/anchor/anchor_utils.py:1]: Maybe we can directly use `utils.py` as the filename under the anchor dir.

File: mmdet/models/anchor_heads/anchor_head.py [hunk 2]
```diff
@@ -120,7 +132,8 @@ def get_anchors(self, featmap_sizes, img_metas, device='cuda'):
             device (torch.device | str): device for returned tensors
 
         Returns:
-            tuple: anchors of each image, valid flags of each image
+            anchor_list: anchors of each image
+            valid_flag_list: valid flags of each image
```
Comment (hellock) [inline mmdet/models/anchor_heads/anchor_head.py:136]:
```python
Returns:
    tuple[list]: xxx
```

File: mmdet/models/anchor_heads/anchor_head.py [hunk 3]
```diff
@@ -84,6 +88,14 @@ def __init__(self,
         self.loss_bbox = build_loss(loss_bbox)
         self.train_cfg = train_cfg
         self.test_cfg = test_cfg
+        if self.train_cfg:
+            self.assigner = build_assigner(self.train_cfg.assigner)
+            # use PseudoSampler when sampling is False
+            if self.sampling and hasattr(self.train_cfg, 'sampler'):
+                sampler_cfg = self.train_cfg.sampler
```
Comment (hellock) [inline mmdet/models/anchor_heads/anchor_head.py:95]: If `self.sampling` is True, but `self.train_cfg` has no field `sampler`, should we raise an exception?

File: mmdet/models/anchor_heads/anchor_head.py [hunk 4]
```diff
@@ -117,10 +129,12 @@ def get_anchors(self, featmap_sizes, img_metas, device='cuda'):
...
             featmap_sizes (list[tuple]): Multi-level feature map sizes.
             img_metas (list[dict]): Image meta info.
-            device (torch.device | str): device for returned tensors
+            device (torch.device | str): Device for returned tensors
 
         Returns:
-            tuple: anchors of each image, valid flags of each image
+            tuple:
+                anchor_list (list[Tensor]): anchors of each image
+                valid_flag_list (list[Tensor]): valid flags of each image
```
Comment (hellock) [inline mmdet/models/anchor_heads/anchor_head.py:137]: valid -> Valid

File: mmdet/models/anchor_heads/anchor_head.py [hunk 5]
```diff
@@ -152,6 +166,177 @@ def get_anchors(self, featmap_sizes, img_metas, device='cuda'):
...
+                      img_metas,
+                      gt_bboxes_ignore_list=None,
+                      gt_labels_list=None,
+                      label_channels=1,
+                      unmap_outputs=True):
+        """Compute regression and classification targets for anchors in
+            multiple images.
+
+        Args:
+            anchor_list (list[list]): Multi level anchors of each image.
```
Comment (hellock) [inline mmdet/models/anchor_heads/anchor_head.py:275]: `anchor_list (list[list[Tensor]]): Multi level anchors of each image. The outer list indicates images, and the inner list corresponds to feature levels of the image. Each element of the inner list is a tensor of shape (num_anchors, 4)`.

File: mmdet/models/anchor_heads/anchor_head.py [hunk 6]
```diff
@@ -152,6 +166,177 @@ def get_anchors(self, featmap_sizes, img_metas, device='cuda'):
...
+                             gt_labels,
+                             img_meta,
+                             label_channels=1,
+                             unmap_outputs=True):
+        """Compute regression and classification targets for anchors in
+            a single image.
+
+        Args:
+            flat_anchors (Tensor): Multi level anchors of the image,
+                shape (num_anchors ,4).
```
Comment (hellock) [inline mmdet/models/anchor_heads/anchor_head.py:183]: Multi-level anchors of the image, which are concatenated into a single tensor of shape (num_anchors ,4).

File: mmdet/models/anchor_heads/anchor_head.py [hunk 7]
```diff
@@ -152,6 +166,183 @@ def get_anchors(self, featmap_sizes, img_metas, device='cuda'):
...
+            labels = unmap(labels, num_total_anchors, inside_flags)
+            label_weights = unmap(label_weights, num_total_anchors,
+                                  inside_flags)
+            bbox_targets = unmap(bbox_targets, num_total_anchors, inside_flags)
+            bbox_weights = unmap(bbox_weights, num_total_anchors, inside_flags)
+
+        return (labels, label_weights, bbox_targets, bbox_weights, pos_inds,
+                neg_inds)
+
+    def get_target(self,
```
Comment (hellock) [inline mmdet/models/anchor_heads/anchor_head.py:263]: `get_targets()`, `_get_targets_single()`, `get_bboxes()`, `_get_bboxes_single()`

File: mmdet/models/anchor_heads/atss_head.py [hunk 8]
```diff
@@ -350,19 +361,19 @@ def get_bboxes_single(self,
...
+                   cfg,
+                   gt_bboxes_ignore_list=None,
+                   gt_labels_list=None,
+                   label_channels=1,
+                   unmap_outputs=True):
         """
-        almost the same with anchor_target, with a little modification,
-        here we need return the anchor
+        almost the same with AnchorHead.get_target, with a little modification,
+            here we need return the anchor
```
Comment (hellock) [inline mmdet/models/anchor_heads/atss_head.py:376]:
```python
"""Get targets for ATSS head.

This method is almost the same as `AnchorHead.get_target()`. Besides returning the targets as the parent method does, it also returns the anchors as the first element of the returned tuple.
"""
```

File: mmdet/models/anchor_heads/guided_anchor_head.py [hunk 9]
```diff
@@ -167,6 +167,9 @@ def __init__(
         self.use_sigmoid_cls = loss_cls.get('use_sigmoid', False)
         self.cls_focal_loss = loss_cls['type'] in ['FocalLoss']
         self.loc_focal_loss = loss_loc['type'] in ['FocalLoss']
+        self.sampling = not self.cls_focal_loss
+        self.ga_sampling = train_cfg is not None and hasattr(
```
Comment (hellock) [inline mmdet/models/anchor_heads/guided_anchor_head.py:171]: Is this necessary?

File: mmdet/models/anchor_heads/reppoints_head.py [hunk 10]
```diff
@@ -383,6 +393,155 @@ def offset_to_pts(self, center_list, pred_list):
...
+                proposals_weights, pos_inds, neg_inds)
+
+    def point_target(self,
+                     proposals_list,
+                     valid_flag_list,
+                     gt_bboxes_list,
+                     img_metas,
+                     gt_bboxes_ignore_list=None,
+                     gt_labels_list=None,
+                     init=True,
```
Comment (hellock) [inline mmdet/models/anchor_heads/reppoints_head.py:468]: `stage='init' | 'refine'` may be better?

File: mmdet/models/anchor_heads/reppoints_head.py [hunk 11]
```diff
@@ -445,17 +604,15 @@ def loss(self,
...
             candidate_list,
             valid_flag_list,
             gt_bboxes,
             img_metas,
-            self.train_cfg.init,
             gt_bboxes_ignore_list=gt_bboxes_ignore,
             gt_labels_list=gt_labels,
-            label_channels=label_channels,
-            background_label=self.background_label,
-            sampling=self.sampling)
```
Comment (hellock) [inline mmdet/models/anchor_heads/reppoints_head.py:458]: How about the argument `sampling`?

File: mmdet/models/bbox_heads/bbox_head.py [hunk 12]
```diff
@@ -79,23 +79,65 @@ def forward(self, x):
         bbox_pred = self.fc_reg(x) if self.with_reg else None
         return cls_score, bbox_pred
 
+    def bbox_target(self,
```
Comment (hellock) [inline mmdet/models/bbox_heads/bbox_head.py:82]: We may remove this method and implement the logic in `get_target()`.

**Pull request merged**